Automattic/wp-calypso
# Handle errors in the theme filter fetch

### Proposed Changes
This _might_ fix the timeout issues on the /themes route! After merging #69538, I dug into the data and found out that the timeout _always_ happens in the fetchThemeFilters middleware.

After I started looking at this code more carefully, I realized that it only ever calls `next` after the filters exist in the store:

```js
const unsubscribe = store.subscribe( () => {
	if ( Object.keys( getThemeFilters( store.getState() ) ).length > 0 ) {
		unsubscribe();
		return next();
	}
} );
```

So if the filters will never exist, this loop will run forever and the middleware will never continue. If some kind of network error happens, the theme filters will never exist -- in other words, we need error handling.

This PR adds that by looking for the "fetch theme filters" notice that gets created by the error handler for the theme filter fetch. 

We then create an error and pass it to `next`. This triggers the error handling behavior in our SSR middleware chain, which effectively means:
1. We get automatically logging of the error
2. Future midleware are skipped
3. The page is still rendered, but not with an SSR layout. The client will have to fetch themes and theme filters.

### Testing Instructions
You can replicate this issue locally on trunk like so:

1. In your hosts file, write `0.0.0.0 public-api.wordpress.com` and comment out any other entries which sandbox the public-api. This will effectively make every request to public-api fail.
4. Run `ping public-api.wordpress.com`. You should see a socket connection error if the above step worked.
5. Run `DEBUG=calypso:themes,calypso:pages,calypso:isomorphic-routing yarn start`
6. In an incognito/private browser window, open `calypso.localhost:3000/themes`
7. The page **should never load**. Wait for like 30 seconds and close the tab  -- verify that you see a performanceMark object logged to the console with `fetchThemeFilters` being the biggest culprit:

<img width="855" alt="image" src="https://user-images.githubusercontent.com/6265975/202795971-0c9bcd86-2890-443b-9c81-1e2f062f2f29.png">

8. Stop the server, checkout this branch, and run `DEBUG=calypso:themes,calypso:pages,calypso:isomorphic-routing yarn start`
9. In an incognito/private browser window, open `calypso.localhost:3000/themes`
10. This time, it might still take maybe 10 seconds for the page to load, but it will load. You should see a message about an error fetching theme filters in the console:

<img width="855" alt="image" src="https://user-images.githubusercontent.com/6265975/202796294-dfcb3840-7bfc-47a5-a502-143017988506.png">

The page should load.

Also note the error handling. You'll see a message like this, which will ultimately show up in logstash once deployed. I encoded the error with JSON.stringify since we can't be completely sure of its type. (Other than that it's not a JS Error object)

```sh
23:08:53.174Z  WARN calypso: Theme fetch failed: {"errno":-61,"code":"ECONNREFUSED","syscall":"connect","address":"0.0.0.0","port":443} (reqId=fb8f4e37-1fbf-4f29-9fc7-543059d87e6f, url=/themes/, env=development, userAgent="Firefox 107", path=/themes/)
    Error: Theme fetch failed: {"errno":-61,"code":"ECONNREFUSED","syscall":"connect","address":"0.0.0.0","port":443}
        at /Users/noahallen/source/wp-calypso/build/webpack:/wp-calypso/client/my-sites/themes/controller.jsx:111:17
        at dispatch (/Users/noahallen/source/wp-calypso/node_modules/redux/lib/redux.js:305:7)
        at next (/Users/noahallen/source/wp-calypso/node_modules/redux-dynamic-middlewares/lib/index.js:17:83)
        at /Users/noahallen/source/wp-calypso/build/webpack:/wp-calypso/client/state/data-layer/wpcom-api-middleware.js:63:11
        at /Users/noahallen/source/wp-calypso/node_modules/redux-thunk/lib/index.js:14:16
        at map (/Users/noahallen/source/wp-calypso/node_modules/redux/lib/redux.js:667:28)
        at Array.map (<anonymous>)
        at next (/Users/noahallen/source/wp-calypso/build/webpack:/wp-calypso/client/state/data-layer/wpcom-http/utils.js:190:25)
        at themeFiltersHandlers (/Users/noahallen/source/wp-calypso/build/webpack:/wp-calypso/client/state/data-layer/wpcom-http/utils.js:116:10)
        at /Users/noahallen/source/wp-calypso/build/webpack:/wp-calypso/client/state/data-layer/wpcom/theme-filters/index.js:45:11
```
